src/main/java/org/truffleruby/language/loader/SourceLoader.java

```diff
@@ -210,13 +213,15 @@ public static Source loadResource(String path) throws IOException {
         }
     }
 
-    private void ensureReadable(String path, File file) {
-        if (!file.exists()) {
-            throw new RaiseException(context.getCoreExceptions().loadError("No such file or directory -- " + path, path, null));
-        }
+    private static void ensureReadable(RubyContext context, String path, File file) {
+        if (context != null) {
```

Comment:
When would the context be null if we're pre-initializing? And, I think we need to standardize on how we want to check for AOT without a context. We have `RubyContext.FIRST_INSTANCE` and that's used by `ParseRopeOperations.isBuildingNativeImage()`. Those might convey intent more clearly here.

Reply:
I think nooping if the context is null can lead to some difficult to debug situations. Maybe throw a different Java-level exception in that case? I'd suggest assertions, but that can drastically increase native image compilation.

Reply:
It's from ParserCache which doesn't have a context.
The exceptions will be thrown as some kind of IOException when reading anyway, so I think this nice early checks are not needed for the ParserCache case.